File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-0.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        globmax, globmin = nums[0], nums[0]
        currmax, currmin = 0, 0
        total = 0

        for n in nums:
            currmax = max(currmax + n, n)
            currmin = min(currmin + n, n)

            total += n

            globmax = max(globmax, currmax)
            globmin = min(globmin, currmin)

        return max(globmax, total - globmin) if globmax > 0 else globmax

File: Data Structures & Algorithms/house-robber/submission-0.cpp
class Solution {
public:
    int dfs(int i,vector<int>&dp,vector<int>&nums){
        int n=nums.size();
        if (i>=n){
            return 0;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        dp[i]=max(nums[i]+dfs(i+2,dp,nums),dfs(i+1,dp,nums));
        return dp[i];

    }
    int rob(vector<int>& nums) {
        int n=nums.size();
        vector<int> dp(n+1,-1);
        return dfs(0,dp,nums);


        
    }
};
